subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/CompilationSourceDirs.java

```diff
@@ -28,63 +31,72 @@
  * Attempts to infer the source root directories for the `source` inputs to a
  * {@link org.gradle.api.tasks.compile.JavaCompile} task, in order to determine the `.class` file that corresponds
  * to any input source file.
- * 
+ *
  * This is a bit of a hack: we'd be better off inspecting the actual source file to determine the name of the class file.
  */
+@NonNullApi
 public class CompilationSourceDirs {
     private static final org.gradle.api.logging.Logger LOG = Logging.getLogger(IncrementalCompilerDecorator.class);
 
-    private final List<Object> sources;
+    private final FileTreeInternal sources;
     private List<File> sourceRoots;
+    private boolean canInferSourceRoots;
 
-    public CompilationSourceDirs(List<Object> sources) {
+    public CompilationSourceDirs(FileTreeInternal sources) {
         this.sources = sources;
     }
 
-    List<File> getSourceRoots() {
+    public List<File> getSourceRoots() {
         if (sourceRoots == null) {
-            sourceRoots = Lists.newArrayList();
-            for (Object source : sources) {
-                if (isDirectory(source)) {
-                    sourceRoots.add((File) source);
-                } else if (isDirectoryTree(source)) {
-                    sourceRoots.add(((DirectoryTree) source).getDir());
-                } else if (isSourceDirectorySet(source)) {
-                    sourceRoots.addAll(((SourceDirectorySet) source).getSrcDirs());
-                } else {
-                    throw new UnsupportedOperationException();
-                }
-            }
+            resolveRoots();
         }
         return sourceRoots;
     }
 
     public boolean canInferSourceRoots() {
-        for (Object source : sources) {
-            if (!canInferSourceRoot(source)) {
-                LOG.info("Cannot infer source root(s) for input with type `{}`. Supported types are `File`, `DirectoryTree` and `SourceDirectorySet`. Unsupported input: {}", source.getClass().getSimpleName(), source);
-                return false;
-            }
+        if (sourceRoots == null) {
+            resolveRoots();
         }
```

Comment:
The check for `null` and setting the value should be in the same method. I'd just let `resolveRoots()` check for `null` and resolve as needed, and then return the resolved value so the caller can work with it.

Reply:
Done.